File: woman-global/mobile/android/app/src/main/java/com/womanglobal/connecther/booking/ProviderGeoSort.kt
package com.womanglobal.connecther.booking

import com.womanglobal.connecther.data.User
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Sorts provider [User] rows by great-circle distance to the seeker when both sides have coordinates.
 * Providers without coordinates are appended after sorted ones (stable relative order).
 */
object ProviderGeoSort {

    /** Earth mean radius in meters (WGS84 approximation). */
    private const val EARTH_RADIUS_M = 6_371_000.0

    fun haversineMeters(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val rLat1 = Math.toRadians(lat1)
        val rLat2 = Math.toRadians(lat2)
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        val a = sin(dLat / 2) * sin(dLat / 2) +
            cos(rLat1) * cos(rLat2) * sin(dLon / 2) * sin(dLon / 2)
        val c = 2 * atan2(sqrt(a), sqrt(1 - a))
        return EARTH_RADIUS_M * c
    }

    /**
     * @param seekerLat seeker latitude, or null to skip distance sort
     * @param seekerLng seeker longitude, or null to skip distance sort
     */
    fun sortByDistance(providers: List<User>, seekerLat: Double?, seekerLng: Double?): List<User> {
        if (providers.isEmpty()) return providers
        if (seekerLat == null || seekerLng == null) return providers

        val withCoords = mutableListOf<Pair<User, Double>>()
        val withoutCoords = mutableListOf<User>()
        for (u in providers) {
            val plat = u.latitude
            val plng = u.longitude
            if (plat != null && plng != null) {
                val m = haversineMeters(seekerLat, seekerLng, plat, plng)
                withCoords.add(u to m)
            } else {
                withoutCoords.add(u)
            }
        }
        val sorted = withCoords.sortedBy { it.second }.map { it.first }
        return sorted + withoutCoords
    }

    fun providerRef(u: User): String = u.user_name.trim().ifBlank { u.id.trim() }
}

File: woman-global/mobile/expo-app/src/utils/providerGeoSort.ts
import type { Provider } from '@/types/models';

const EARTH_RADIUS_M = 6_371_000;

export function haversineMeters(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const rLat1 = (lat1 * Math.PI) / 180;
  const rLat2 = (lat2 * Math.PI) / 180;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(rLat1) * Math.cos(rLat2) * Math.sin(dLon / 2) * Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_M * c;
}

/** Stable sort: closest first when seeker coords exist; providers without coords follow (Kotlin `ProviderGeoSort`). */
export function sortProvidersByDistance(
  providers: Provider[],
  seekerLat: number | null | undefined,
  seekerLng: number | null | undefined,
): Provider[] {
  if (providers.length === 0) return providers;
  if (seekerLat == null || seekerLng == null || !Number.isFinite(seekerLat) || !Number.isFinite(seekerLng)) {
    return [...providers];
  }

  const withCoords: { p: Provider; m: number }[] = [];
  const without: Provider[] = [];
  for (const p of providers) {
    const plat = p.latitude;
    const plng = p.longitude;
    if (plat != null && plng != null && Number.isFinite(plat) && Number.isFinite(plng)) {
      withCoords.push({ p, m: haversineMeters(seekerLat, seekerLng, plat, plng) });
    } else {
      without.push(p);
    }
  }
  withCoords.sort((a, b) => a.m - b.m);
  return [...withCoords.map((x) => x.p), ...without];
}

export function providerRefForSort(p: Provider): string {
  return (p.user_name ?? '').trim() || String(p.id);
}
